src/_guides/language/evolution.md

```diff
@@ -127,25 +151,99 @@ void main() {
 }
 ```
 
-## Dart 2.8
+### Dart 2.8
 
 Dart 2.8 didn't add any features to the Dart language, but it did
 contain a number of preparatory [breaking changes][2.8 breaking changes] to ensure great
 nullability-related usability and performance in the upcoming
-null safety feature.
+[null safety][] feature.
 
 It also contained a faster pub tool, and a new [pub outdated][] command.
 
-## Dart 2.9
+### Dart 2.9
 
 Dart 2.9 didn't add any features to the Dart language.
 
-## Dart 2.10
+### Dart 2.10
 
 Dart 2.10 didn't add any features to the Dart language,
 but it added an expanded [`dart` tool][dart-tool] that's 
 analogous to the Flutter SDK's `flutter` tool.
 
+## Language versioning
+
+Before language versioning,
+Dart tools assumed that all Dart code could use all features
+that were supported by the tool's SDK version.
+This assumption causes problems when a package
+that uses a _newer_ language feature
+has [SDK constraints][] that are too low.
+For example, consider a package that uses set literals (2.2)
+but has the SDK constraints `">=2.0.0 <3.0.0"`.
+People using Dart 2.0.0 or 2.1.0 can download the package,
+but their tools can't understand the set literals.
+
+With language versioning,
+a single Dart SDK simultaneously supports
+multiple different versions of the Dart language.
+When you compile some Dart code,
+Dart figures out what version the code is targeting,
+and you're warned if the code uses a feature
+that was introduced after the version in the lower SDK constraint.
+
+On rare occasions, Dart introduces a
+backwards-incompatible feature like [null safety][].
+Previously valid code might not work with null safety,
+so to get the benefits of null safety you'll need to migrate your code.
+And because migrating to null safety can't be completely automated,
+you'll need to be able to migrate one library
+(usually one Dart file) at a time.
+
+{% comment %}
+  Once /null-safety/migrating-to-null-safety exists,
+  the previous paragraph should link to it.
+{% endcomment %}
+
+To enable file-by-file migration,
+Dart 2.10 introduced **per-library language version selection**.
+**[PENDING: was it really a 2.10 feature?]**
```

Comment:
I could be wrong, but I think language versioning shipped in 2.7.0. (@lrhn would know more.) We didn't announce it because it didn't really do anything yet (except for give you warnings if you tried to use extension methods in packages whose SDK constraints were too low).

Reply:
I *think* it shipped in 2.8. (That's why `package:package_config` defaults code with only a `.packages` file to version 2.7, the last version prior to language versioning existing. From 2.8 you should stop using `.packages`.)
There is no entry in our changelog for 2.8.0. We go directly to 2.8.1 and states that there are no new language features.
The only mention of "language versioning" in the CHANGELOG.md is when pub starts generating the `package_config.json` file in 2.7, which it did prior to launching language versioning). There is a link to the package versioning spec in the 2.8.1 notes, in a fix to Pub.

Which also means that it's shouldn't be affecting extension methods, which were released in 2.7 (and enabled for preview in 2.6).

I think language versioning snuck under the radar because it was never an *experiment*. It's just a compiler feature.